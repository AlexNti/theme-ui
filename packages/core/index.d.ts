import React from 'react';
import { Theme } from './theme';
export declare const jsx: typeof React.createElement;
export declare const Context: React.Context<{
    __EMOTION_VERSION__: string;
    theme: Theme;
}>;
export declare const useThemeUI: () => {
    __EMOTION_VERSION__: string;
    theme: Theme;
};
export declare const merge: {
    (a: any, b: any): unknown;
    all(...args: any[]): object;
};
export interface ThemeProviderProps {
    theme: Partial<Theme> | ((outerTheme: Theme) => Theme);
    children?: React.ReactNode;
}
export declare function ThemeProvider({ theme, children }: ThemeProviderProps): React.FunctionComponentElement<{
    context: any;
    children: any;
}>;
