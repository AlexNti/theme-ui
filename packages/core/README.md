
# @theme-ui/core

Core Emotion and JSX implementation for Theme UI

## API

- `jsx`
- `Context`
- `useThemeUI`
- `merge`
- `ThemeProvider`
