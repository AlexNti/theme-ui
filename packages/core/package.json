{
  "name": "@theme-ui/core",
  "version": "0.2.48",
  "main": "dist/index.js",
  "module": "dist/index.esm.js",
  "scripts": {
    "prepare": "microbundle --no-compress",
    "watch": "microbundle watch --no-compress"
  },
  "repository": "system-ui/theme-ui",
  "author": "Brent Jackson",
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@emotion/core": "^10.0.22"
  },
  "devDependencies": {
    "react": "^16.11.0"
  }
}
