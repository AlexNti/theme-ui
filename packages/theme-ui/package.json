{
  "name": "theme-ui",
  "version": "0.0.17",
  "description": "Build consistent, themeable React UIs based on design system constraints and design tokens",
  "main": "dist/index.js",
  "module": "dist/index.esm.js",
  "author": "Brent Jackson <jxnblk@gmail.com>",
  "license": "MIT",
  "repository": "system-ui/theme-ui",
  "scripts": {
    "prepare": "microbundle --no-compress",
    "watch": "microbundle watch --no-compress",
    "test": "jest"
  },
  "peerDependencies": {
    "@emotion/core": "^10.0.0",
    "@emotion/styled": "^10.0.0",
    "@mdx-js/react": "^1.0.0"
  },
  "dependencies": {
    "@styled-system/css": "^2.0.4",
    "lodash.merge": "^4.6.0",
    "react": "^16.8.0",
    "styled-system": "^4.0.0"
  },
  "devDependencies": {
    "@emotion/core": "^10.0.0",
    "@emotion/styled": "^10.0.0",
    "@mdx-js/react": "^1.0.0",
    "@babel/core": "^7.4.3",
    "@babel/preset-env": "^7.4.3",
    "@mdx-js/mdx": "^1.0.0",
    "@testing-library/react": "^8.0.1",
    "jest": "^24.7.0",
    "jest-emotion": "^10.0.10",
    "microbundle": "^0.11.0",
    "react-test-renderer": "^16.8.6"
  },
  "jest": {
    "testRegex": "./test/.*.js$",
    "coverageReporters": [
      "lcov",
      "text",
      "html"
    ],
    "coverageThreshold": {
      "global": {
        "branches": 90,
        "functions": 90,
        "lines": 90,
        "statements": 90
      }
    }
  }
}
