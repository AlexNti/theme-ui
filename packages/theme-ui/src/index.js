// TODO: rollup/microbundle screw this up

export { ThemeProvider } from './provider'
export { BaseStyles } from './base-styles'
export {
  css,
  get,
} from '@theme-ui/css'
export {
  jsx,
  Context,
  merge,
  useThemeUI,
} from '@theme-ui/core'
export {
  useColorMode,
  InitializeColorMode,
  ColorMode,
} from '@theme-ui/color-modes'
export {
  Styled,
  components,
} from '@theme-ui/mdx'
export * from '@theme-ui/components'

/*
export { ColorMode, useColorMode, InitializeColorMode } from './color-modes'
export { Box, Flex, Layout, Header, Main, Container, Footer } from './layout'
export { BaseStyles } from './base-styles'
export { css, get } from '@theme-ui/css'
*/
