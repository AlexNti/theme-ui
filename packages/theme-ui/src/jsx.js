import { jsx as emotion } from '@emotion/core'
import css from '@styled-system/css'

const getCSS = props => theme => {
  const styles = css(props.sx)(theme)
  const raw = typeof props.css === 'function' ? props.css(theme) : props.css
  return [styles, raw]
}

export const jsx = (type, props, ...children) =>
  emotion.apply(undefined, [
    type,
    props
      ? {
          ...props,
          css: getCSS(props),
        }
      : null,
    ...children,
  ])

export default jsx
