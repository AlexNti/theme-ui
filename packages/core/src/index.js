// todo/figure out
// - [ ] color modes
// - [ ] custom properties
// - [ ] mdx components @theme-ui/mdx
// - [ ] better default color mode name API

// API
//  <ThemeProvider theme={theme}>
//    <ColorModeProvider>
//      <MDXProvider>
//        <ColorMode />
//        <App />
//      </MDXProvider>
//    </ColorModeProvider>
//  </ThemeProvider>

import {
  jsx as emotion,
  ThemeContext as EmotionContext,
} from '@emotion/core'
import { css, get } from '@theme-ui/css'
import {
  createContext,
  useContext,
  useState,
  useEffect,
} from 'react'
import deepmerge from 'deepmerge'
import { version as __EMOTION_VERSION__ } from '@emotion/core/package.json'

const getCSS = props => {
  if (!props.sx && !props.css) return undefined
  return theme => {
    const styles = css(props.sx)(theme)
    const raw = typeof props.css === 'function' ? props.css(theme) : props.css
    return [styles, raw]
  }
}

const parseProps = props => {
  if (!props) return null
  const next = {}
  for (let key in props) {
    if (key === 'sx') continue
    next[key] = props[key]
  }
  const css = getCSS(props)
  if (css) next.css = css
  return next
}

export const jsx = (type, props, ...children) =>
  emotion.apply(undefined, [type, parseProps(props), ...children])

export const Context = createContext({
  __EMOTION_VERSION__,
  theme: null,
})

export const useThemeUI = () => useContext(Context)

const canUseSymbol = typeof Symbol === 'function' && Symbol.for

const REACT_ELEMENT = canUseSymbol ? Symbol.for('react.element') : 0xeac7
const FORWARD_REF = canUseSymbol ? Symbol.for('react.forward_ref') : 0xeac7

const isMergeableObject = n => {
  return (
    !!n &&
    typeof n === 'object' &&
    n.$$typeof !== REACT_ELEMENT &&
    n.$$typeof !== FORWARD_REF
  )
}

const arrayMerge = (destinationArray, sourceArray, options) => sourceArray

export const merge = (a, b) =>
  deepmerge(a, b, { isMergeableObject, arrayMerge })

merge.all = (...args) => deepmerge.all(args, { isMergeableObject, arrayMerge })

const BaseProvider = ({ context, children }) =>
  jsx(
    EmotionContext.Provider, { value: context.theme },
    jsx(Context.Provider, {
      value: context,
      children
    })
  )

// possible new api
//  <ThemeProvider theme={theme}>
//  <ColorModeProvider>
//  <ComponentsProvider components={components}>

export const ThemeProvider = ({
  theme,
  children
}) => {
  const outer = useThemeUI()
  const context = {
    ...outer,
    theme: typeof theme === 'function'
      ? theme(outer.theme)
      : merge.all({}, outer.theme, theme)
  }
  return jsx(BaseProvider, {
    context,
    children
  })
}

// color modes
const STORAGE_KEY = 'theme-ui-color-mode'
const storage = {
  get: init => window.localStorage.getItem(STORAGE_KEY) || init,
  set: value => window.localStorage.setItem(STORAGE_KEY, value),
}

export const useColorState = theme => {
  const [mode, setMode] = useState(theme.initialColorModeName || 'default')

  // initialize state
  useEffect(() => {
    const stored = storage.get()
    document.body.classList.remove('theme-ui-' + stored)
    // consider prefers-color-scheme media query
    if (!stored || stored === mode) return
    setMode(stored)
  }, [])

  useEffect(() => {
    if (!mode) return
    storage.set(mode)
  }, [mode])

  if (process.env.NODE_ENV !== 'production') {
    if (
      theme.colors &&
      theme.colors.modes &&
      Object.keys(theme.colors.modes).indexOf(theme.initialColorModeName) > -1
    ) {
      console.warn(
        'The `initialColorModeName` value should be a unique name' +
          ' and cannot reference a key in `theme.colors.modes`.'
      )
    }
  }

  return [mode, setMode]
}

export const useColorMode = () => {
  const { colorMode, setColorMode } = useThemeUI()

  if (typeof setColorMode !== 'function') {
    throw new Error(`[useColorMode] requires the ColorModeProvider component`)
  }

  return [colorMode, setColorMode]
}

const applyColorMode = (theme, mode) => {
  if (!mode) return theme
  const modes = get(theme, 'colors.modes', {})
  return merge.all({}, theme, {
    colors: get(modes, mode, {}),
  })
}

export const ColorModeProvider = ({
  children,
}) => {
  const outer = useThemeUI()
  const [colorMode, setColorMode] = useColorModeState(outer.theme)
  const theme = applyColorMode(outer.theme, colorMode)
  const context = {
    ...outer,
    theme,
    colorMode,
    setColorMode,
  }

  return jsx(BaseProvider, {
    context,
    children,
  })
}

// export const ColorMode
// export const InitializeColorMode
