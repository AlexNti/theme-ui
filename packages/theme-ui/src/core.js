// refactored stateful theme provider
import jsx from './jsx'
import {
  useContext,
  useState,
  useReducer,
  useEffect,
  useLayoutEffect,
} from 'react'
import styled from '@emotion/styled'
import { ThemeContext as Emotion, Global } from '@emotion/core'
import { MDXProvider } from '@mdx-js/react'
import { css, get } from '@styled-system/css'
import merge from 'lodash.merge'
import themed from './themed'
import { Context } from './context'

const IS_DEV = process.env.NODE_ENV !== 'production'

const reducer = (state, next) => merge({}, state, next)

const Provider = ({ context, children }) => {
  const [ state, setTheme ] = useReducer(reducer, context.theme)
  const theme = applyColorMode(state, context.colorMode)
  theme.setTheme = setTheme

  useEffect(() => {
    // todo: support multiple instances
    window.THEME_UI = window.THEME_UI || theme
  }, [])

  return jsx(Emotion.Provider, { value: theme },
    jsx(MDXProvider, { components: context.components },
      jsx(Context.Provider, { value: context },
        children
      )
    )
  )
}

const createComponents = (components = {}) => {
  const next = {}
  Object.keys(components).forEach(key => {
    next[key] = styled(components[key])(themed(key))
  })
  return next
}

const applyColorMode = (theme = {}, colorMode) => ({
  ...theme,
  colors: merge({},
    theme.colors,
    get(theme, `colors.modes.${colorMode}`, {})
  )
})

const STORAGE_KEY = 'theme-ui-color-mode'

const storage = {
  get: (init) => window.localStorage.getItem(STORAGE_KEY) || init,
  set: (value) => window.localStorage.setItem(STORAGE_KEY, value),
}

const getMediaQuery = () => {
  const darkQuery = '(prefers-color-scheme: dark)'
  const mql = window.matchMedia ? window.matchMedia(darkQuery) : {}
  const dark = mql.media === darkQuery
  return dark && mql.matches
}

const useStoredColorMode = ({ colorMode, setColorMode }) => {
  useLayoutEffect(() => {
    const stored = storage.get()
    document.body.classList.remove('theme-ui-' + stored)
    if (!stored) {
      const dark = getMediaQuery()
      if (dark) setColorMode('dark')
      return
    }
    setColorMode(stored)
  }, [])
}

const useThemeContext = ({ theme, components }) => {
  const outer = useContext(Context)
  const isNested = !!outer.colorMode

  const [ colorMode, setColorMode ] = useState(theme.initialColorMode)

  const context = merge({
    colorMode,
    setColorMode,
  }, outer, {
    theme,
    components: createComponents(components)
  })

  useStoredColorMode(context)

  return context
}

export const ThemeProvider = ({
  theme = {},
  components,
  children,
}) => {
  const context = useThemeContext({ theme, components })

  return jsx(Provider, { context, children })
}

export const useColorMode = (initialMode) => {
  const { colorMode, setColorMode } = useContext(Context)

  if (typeof setColorMode !== 'function') {
    throw new Error(
      `[useColorMode] requires the ThemeProvider component`
    )
  }

  return [ colorMode, setColorMode ]
}

const bodyColor = (theme = {}) => {
  if (!theme.colors || !theme.colors.modes) return
  const { modes } = theme.colors
  const styles = {}
  Object.keys(modes).forEach(mode => {
    const colors = modes[mode]
    styles[`&.theme-ui-${mode}`] = {
      color: colors.text,
      bg: colors.background,
    }
  })

  return css({
    body: {
      ...styles,
      color: 'text',
      bg: 'background',
    },
  })(theme)
}

export const ColorMode = () => {
  const { colorMode } = useContext(Context)
  useEffect(() => {
    if (!colorMode) return
    storage.set(colorMode)
  }, [ colorMode ])

  return jsx(Global, { styles: bodyColor })
}

