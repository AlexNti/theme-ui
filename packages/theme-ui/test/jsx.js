import { transformSync } from '@babel/core'
import jsx from '../src/jsx'

const fixture = `/** @jsx jsx */
import { jsx } from 'theme-ui'

export default props =>
  <div
    {...props}
    css={{
      m: 0,
      bg: 'primary',
    }}
  />
`

test('converts css prop to @styled-system/css call', () => {
  const { code } = transformSync(fixture, {
    presets: [
      '@babel/preset-react',
    ],
    plugins: [
      '@babel/plugin-transform-runtime',
    ],
  })
  expect(code).toMatchSnapshot()
})

test('works with emotion preset', () => {
  const { code } = transformSync(fixture, {
    presets: [
      '@emotion/babel-preset-css-prop',
      '@babel/preset-react',
    ],
    plugins: [
      '@babel/plugin-transform-runtime',
    ],
  })
  expect(code).toMatch(/background-color:primary;/)
  expect(code).toMatch(/margin:0;/)
  // expect(code).toMatchSnapshot()
})
