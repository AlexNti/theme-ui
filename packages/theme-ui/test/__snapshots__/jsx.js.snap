// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converts css prop to @styled-system/css call 1`] = `
"\\"use strict\\";

var _interopRequireDefault = require(\\"@babel/runtime/helpers/interopRequireDefault\\");

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports[\\"default\\"] = void 0;

var _extends2 = _interopRequireDefault(require(\\"@babel/runtime/helpers/extends\\"));

var _themeUi = require(\\"theme-ui\\");

/** @jsx jsx */
var _default = function _default(props) {
  return (0, _themeUi.jsx)(\\"div\\", (0, _extends2[\\"default\\"])({}, props, {
    css: {
      m: 0,
      bg: 'primary'
    }
  }));
};

exports[\\"default\\"] = _default;"
`;

exports[`works with emotion preset 1`] = `
"\\"use strict\\";

var _interopRequireDefault = require(\\"@babel/runtime/helpers/interopRequireDefault\\");

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports[\\"default\\"] = void 0;

var _core = require(\\"@emotion/core\\");

var _extends2 = _interopRequireDefault(require(\\"@babel/runtime/helpers/extends\\"));

var _themeUi = require(\\"theme-ui\\");

/** @jsx jsx */
var _ref = process.env.NODE_ENV === \\"production\\" ? {
  name: \\"4ep6w8\\",
  styles: \\"m:0;bg:primary;\\"
} : {
  name: \\"4ep6w8\\",
  styles: \\"m:0;bg:primary;\\",
  map: \\"/*# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbInVua25vd24iXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IkFBTUkiLCJmaWxlIjoidW5rbm93biIsInNvdXJjZXNDb250ZW50IjpbIi8qKiBAanN4IGpzeCAqL1xuaW1wb3J0IHsganN4IH0gZnJvbSAndGhlbWUtdWknXG5cbmV4cG9ydCBkZWZhdWx0IHByb3BzID0+XG4gIDxkaXZcbiAgICB7Li4ucHJvcHN9XG4gICAgY3NzPXt7XG4gICAgICBtOiAwLFxuICAgICAgYmc6ICdwcmltYXJ5JyxcbiAgICB9fVxuICAvPlxuIl19 */\\"
};

var _default = function _default(props) {
  return (0, _themeUi.jsx)(\\"div\\", (0, _extends2[\\"default\\"])({}, props, {
    css: _ref
  }));
};

exports[\\"default\\"] = _default;"
`;
